core/src/main/scala/org/apache/spark/storage/StorageUtils.scala

```diff
@@ -118,8 +140,20 @@ class StorageStatus(val blockManagerId: BlockManagerId, val maxMem: Long) {
         } else {
           None
         }
+      case BroadcastBlockId(broadcastId, _) =>
+        // Actually remove the block, if it exists
+        if (_broadcastBlocks.contains(broadcastId)) {
+          val removed = _broadcastBlocks(broadcastId).remove(blockId)
+          // If the given RDD has no more blocks left, remove the RDD
```

Comment:
you could tighten it up even more if you moved that helper method to an extractor:

```
object RddOrBlockId {
  def unapply(blockId: BlockId): Option[(Int, Map[BlockId,BlockStatus])] = {
    case RDDBlockId(rddId, _) => Some((rddId, _rddBlocks))
    case BroadcastBlockId(broadcastId, _) => Some((broadcastId, _broadcastBlocks))
    case _ => None
  }
}
```

then the actual matching in each of these blocks would simplify to:

```
  case RddOrBlockId(id, blockMap) =>
    val removed = blockMap(id).remove(blockId)
    ...
```
(the point is, the helper function to get the id & blockMap goes right into the pattern match.)  But maybe this is a bit of scala that is too esoteric for our code base for saving one line in each of these cases.